File: #99_Recover_Binary_Search_Tree/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    int bug1, bug2;
    int count;
    int prev;
    public void recoverTree(TreeNode root) {
		if (root == null) return;
		bug1 = bug2 = 0;
		count = 0;
		prev = Integer.MIN_VALUE;
		inorder(root);
		// System.out.println(bug1 + ", " + bug2);
		inorderFix(root);
    }

    private void inorder(TreeNode node) {
    	if (node == null) return;
    	inorder(node.left);
    	if (node.val < prev) {
    		if (count == 0) {
    			// System.out.println("update bug1");
    			bug1 = prev;
    			bug2 = node.val;
    			count++;
    		}
    		else {
    			// System.out.println("update bug2");
    			bug2 = node.val;
    		}
    	}
    	prev = node.val;
    	inorder(node.right);
    }

    private void inorderFix(TreeNode node) {
    	if (node == null) return;
    	inorderFix(node.left);
    	if (node.val == bug1) {
    		node.val = bug2;
    	}
    	else if (node.val == bug2) {
    		node.val = bug1;
    	}
    	inorderFix(node.right);
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	TreeNode head = new TreeNode(0);
    	TreeNode node = new TreeNode(1);
    	head.left = node;
    	sol.recoverTree(head);
    }
}

File: #99_Recover_Binary_Search_Tree/Solution.py
'''
O(1) space using Morris traversal
'''

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:
	# @param {TreeNode} root
	# @return {void} Do not return anything, modify root in-place instead.
	def recoverTree(self, root):
		badNode0, badNode1 = None, None
		prev, cur = None, root
		while cur:
			rm = self.findRightMostNode(cur) # make sure node is not None
			if not rm or rm.right is cur:
				if prev and prev.val > cur.val:
					if badNode0 is None:
						badNode0 = prev
						badNode1 = cur # tricky
					else:
						badNode1 = cur
				if rm:
					rm.right = None # break link
				prev = cur
				cur = cur.right
			else: # visit left
				rm.right = cur # build link
				cur = cur.left
		if badNode0 and badNode1:
			tmp = badNode0.val
			badNode0.val = badNode1.val
			badNode1.val = tmp

	# find the right most node in the left subtree
	def findRightMostNode(self, root):
		node = root.left
		while node and node.right:
			if node.right is root: # kill infinite loop
				break
			node = node.right
		return node

sol = Solution()
root = TreeNode(0)
root.left = TreeNode(1)
sol.recoverTree(root)